### ISSUE TITLE
[Cupertino] Architecture for Cupertino focus halos and Full Keyboard Access (FKA)

### ISSUE THREAD

**Issue by O-Hannonen** [flutter/flutter#184305]
### Background
I originally opened PR #179225 to add keyboard focus support to `CupertinoContextMenu` and `CupertinoContextMenuAction`. The PR sparked more discussion on how Flutters basic keyboard traversal and iOS Full Keyboard Access (FKA) should split responsibilities and interact with each other. Let's continue that discussion in this issue. 

### The Core Problem: Handling iOS Focus Halos
The main blocker is figuring out how Flutter interacts with Apple's Full Keyboard Access (FKA) feature. The OS handles focus differently depending on the user's settings, which leads to two distinct scenarios:

**Scenario 1: Full Keyboard Access (FKA) is ON**
iOS handles keyboard navigation automatically and draws its own system-level focus halos. If Flutter draws its own `CupertinoFocusHalo`, the system and the framework fight each other, potentially causing double halos or broken navigation.

**Scenario 2: Full Keyboard Access (FKA) is OFF**
Flutter still supports basic keyboard traversal. Because the OS does not draw focus rings in this state, Flutter must render its own `CupertinoFocusHalo` to provide visual feedback to the user.

### Questions to Resolve
* How do we handle drawing the halo? Do we conditionally disable `CupertinoFocusHalo` when FKA is enabled?
* Can the Flutter engine detect the iOS FKA state and expose it to the framework?
* How do we map Flutter's internal focus nodes to iOS accessibility elements so native FKA can highlight them correctly? Eg. how do we communicate the shapes to native side so not all focusable items have square borders when FKA is ON?

**Comment by LongCatIsLooong** [comment]
> is it possible to always hide the FKA's system halo? if not, I can't think of a way to avoid double halo problem

I don't remember if this worked for me but there's an API to change/hide the effect: https://developer.apple.com/documentation/uikit/uifocusitem/focuseffect

**Comment by LongCatIsLooong** [comment]
Interestingly on macos app developers [can tell whether FKA is enabled](https://developer.apple.com/documentation/appkit/nsapplication/isfullkeyboardaccessenabled) (but I don't know if there's a way to monitor it). But on iOS I still can find API that allows us to do that. 

@O-Hannonen 
Instead of resorting to a private API I think the better alternative would be hiding the focus effect of the SemanticsObjects and roughly at the same time land https://github.com/flutter/flutter/pull/179225 (is it still active?) so app users still get visual feedback about the focus.

**Comment by chunhtai** [comment]
reopen this since the discussion should be resolved in between flutter team

**Comment by chunhtai** [comment]
@LongCatIsLooong correct me if i am wrong. I think our last discussion is that when semantics is enabled, all focusable widget should draw cupertino focus halos around it (maybe every focus widget?) and then disable system focus halo. Is this correct?

**Comment by LongCatIsLooong** [comment]
Yeah we can disable the iOS system focus effect but it's difficult to determine whether FKA is enabled. So it will be easier if the framework draws the halos all the time for the control that currently has the primary focus.

**Comment by chunhtai** [comment]
> it will be easier if the framework draws the halos all the time for the control that currently has the primary focus.

I assume we can draw it only when a11y is enabled?
